File: 0004/python3/solution.py
"""
A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
"""


def palindromify(x, y, z):
    return int(str(z) + str(y) + \
               str(x) + str(x) + \
               str(y) + str(z))

# Brute force :(
for one in range(9, 0, -1):
    for two in range(9, -1, -1):
        for three in range(9, -1, -1):
            num = palindromify(three, two, one)
            for a in range(100, 1000):
                for b in range(100, 1000):
                    if a * b == num:
                        print("{} * {} = {}".format(str(a), str(b), str(a * b)))
                        exit(0)

File: 0004/go/solution.go
/*
A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
*/

package main

import "fmt"
import "os"

func palindromify(x int, y int, z int) int {
     return (1e5 * z + 1e4 * y + 1e3 * x +
     	    1e2 * x + 1e1 * y + z);
}

/* Brute force :( */
func main() {
     for one := 9; one >= 1; one -= 1 {
     	 for two := 9; two >= 0; two -= 1 {
	     for three := 9; three >= 0; three -= 1 {
	         var num = palindromify(three, two, one)
	         for a := 100; a < 1000; a += 1 {
	     	     for b := 100; b < 1000; b += 1 {
		         if a * b == num {
			     fmt.Println(num)
			     os.Exit(0)
			 }
		     }
	         }
	     }
	 }
     }
}